HQM crashes world when book is closed or quest is chosen

I have been working on my HQM book for about 21 hours and out of the blue every time I close the book i get "A fatal error has occurred, this connection is terminated". It doesn't crash the client it just crashes the world. It worked before I added a chapter dealing with ProjectE so I tried removing the ProjectE chapter but it still crashes the world. If I try clicking on the first quest in chapter 1 the world crashes. I have over 98 quests typed out and configured. I really don't want to lose it.



12:34:50 | FML | There was a critical exception handling a packet on channel hcQuesting io.netty.handler.codec.EncoderException: java.lang.IllegalArgumentException: The string is too long for this encoding. 	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:106) 	at io.netty.handler.codec.MessageToMessageCodec.write(MessageToMessageCodec.java:116) 	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738) 	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:801) 	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:814) 	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:794) 	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:831) 	at io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:1032) 	at io.netty.channel.AbstractChannel.writeAndFlush(AbstractChannel.java:296) 	at net.minecraftforge.fml.common.network.simpleimpl.SimpleNetworkWrapper.sendTo(SimpleNetworkWrapper.java:250) 	at hardcorequesting.network.NetworkManager.sendToPlayer(NetworkManager.java:47) 	at hardcorequesting.quests.QuestLine.sendServerSync(QuestLine.java:86) 	at hardcorequesting.event.PlayerTracker.onPlayerLogin(PlayerTracker.java:39) 	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler_1988_PlayerTracker_onPlayerLogin_PlayerLoggedInEvent.invoke(.dynamic) 	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler.invoke(ASMEventHandler.java:90) 	at net.minecraftforge.fml.common.eventhandler.EventBus.post(EventBus.java:182) 	at net.minecraftforge.fml.common.FMLCommonHandler.firePlayerLoggedIn(FMLCommonHandler.java:575) 	at net.minecraft.server.management.PlayerList.initializeConnectionToPlayer(PlayerList.java:219) 	at shadows.fastbench.net.HijackedPlayerList.initializeConnectionToPlayer(HijackedPlayerList.java:19) 	at net.minecraftforge.fml.common.network.handshake.NetworkDispatcher.completeServerSideConnection(NetworkDispatcher.java:255) 	at net.minecraftforge.fml.common.network.handshake.NetworkDispatcher.access$100(NetworkDispatcher.java:72) 	at net.minecraftforge.fml.common.network.handshake.NetworkDispatcher$1.func_73660_a(NetworkDispatcher.java:205) 	at net.minecraft.network.NetworkManager.func_74428_b(NetworkManager.java:285) 	at net.minecraft.network.NetworkSystem.func_151269_c(NetworkSystem.java:180) 	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:790) 	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:668) 	at net.minecraft.server.integrated.IntegratedServer.func_71217_p(IntegratedServer.java:252) 	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:526) 	at java.lang.Thread.run(Thread.java:745) Caused by: java.lang.IllegalArgumentException: The string is too long for this encoding. 	at org.apache.commons.lang3.Validate.isTrue(Validate.java:158) 	at net.minecraftforge.fml.common.network.ByteBufUtils.writeUTF8String(ByteBufUtils.java:166) 	at hardcorequesting.network.message.PlayerDataSyncMessage.toBytes(PlayerDataSyncMessage.java:53) 	at net.minecraftforge.fml.common.network.simpleimpl.SimpleIndexedCodec.encodeInto(SimpleIndexedCodec.java:30) 	at net.minecraftforge.fml.common.network.simpleimpl.SimpleIndexedCodec.encodeInto(SimpleIndexedCodec.java:26) 	at net.minecraftforge.fml.common.network.FMLIndexedMessageToMessageCodec.encode(FMLIndexedMessageToMessageCodec.java:81) 	at io.netty.handler.codec.MessageToMessageCodec$1.encode(MessageToMessageCodec.java:67) 	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:88) 	... 28 more
-- | -- | --
12:34:50 | FML | Network Disconnect: A fatal error has occurred, this connection is terminated
12:34:50 | net.minecraft.network.NetHandlerPlayServer | PigguJones lost connection: A fatal error has occurred, this connection is terminated

--- a/src/main/java/hardcorequesting/network/message/PlayerDataSyncMessage.java
+++ b/src/main/java/hardcorequesting/network/message/PlayerDataSyncMessage.java
@@ -14,14 +14,28 @@ import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;
 
 import java.io.IOException;
 import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.List;
 
 public class PlayerDataSyncMessage implements IMessage {
 
     private boolean local, serverWorld, questing, hardcore;
-    private String team, data;
+    private String team;
+
+    private List<String> data = new ArrayList<String>();
 
     public PlayerDataSyncMessage() {
 
+    }
+
+    private List<String> SplitData (String input, int size) {
+        List<String> output = new ArrayList<String>();
+        int len = input.length();
+        for (int i=0; i<len; i+= size) {
+            output.add(input.substring(i, Math.min(len, i+size)));
+        }
+
+        return output;
     }
 
     public PlayerDataSyncMessage(boolean local, boolean serverWorld, EntityPlayer player) {
@@ -30,7 +44,8 @@ public class PlayerDataSyncMessage implements IMessage {
         this.questing = QuestingData.isQuestActive();
         this.hardcore = QuestingData.isHardcoreActive();
         this.team = Team.saveTeam(player);
-        this.data = QuestingData.saveQuestingData(player);
+        String stringData = QuestingData.saveQuestingData(player);
+        this.data = SplitData(stringData, 3000);
     }
 
     @Override
@@ -40,7 +55,11 @@ public class PlayerDataSyncMessage implements IMessage {
         this.questing = buf.readBoolean();
         this.hardcore = buf.readBoolean();
         this.team = ByteBufUtils.readUTF8String(buf);
-        this.data = ByteBufUtils.readUTF8String(buf);
+        int count = buf.readInt();
+        data.clear();
+        for (int i = 0; i<count; i++) {
+            data.add(ByteBufUtils.readUTF8String(buf));
+        }
     }
 
     @Override
@@ -50,7 +69,10 @@ public class PlayerDataSyncMessage implements IMessage {
         buf.writeBoolean(this.questing);
         buf.writeBoolean(this.hardcore);
         ByteBufUtils.writeUTF8String(buf, team);
-        ByteBufUtils.writeUTF8String(buf, data);
+        buf.writeInt(this.data.size());
+        for (String val : this.data) {
+            ByteBufUtils.writeUTF8String(buf, val);
+        }
     }
 
     public static class Handler implements IMessageHandler<PlayerDataSyncMessage, IMessage> {
@@ -72,7 +94,7 @@ public class PlayerDataSyncMessage implements IMessage {
                 }
                 try (PrintWriter out = new PrintWriter(SaveHandler.getRemoteFile("data"))) {
                     out.print("[");
-                    out.print(message.data);
+                    out.print(String.join("", message.data));
                     out.print("]");
                 }
                 try (PrintWriter out = new PrintWriter(SaveHandler.getRemoteFile("state"))) {

--- a/src/main/java/hardcorequesting/network/message/TeamUpdateMessage.java
+++ b/src/main/java/hardcorequesting/network/message/TeamUpdateMessage.java
@@ -10,29 +10,50 @@ import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
 import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
 import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;
 
+import java.util.ArrayList;
+import java.util.List;
+
 public class TeamUpdateMessage implements IMessage {
 
     private TeamUpdateType type;
-    private String data;
+    private List<String> data = new ArrayList<String>();
 
     public TeamUpdateMessage() {
     }
 
+    private List<String> SplitData (String input, int size) {
+        List<String> output = new ArrayList<String>();
+        int len = input.length();
+        for (int i=0; i<len; i+= size) {
+            output.add(input.substring(i, Math.min(len, i+size)));
+        }
+
+        return output;
+    }
+
     public TeamUpdateMessage(TeamUpdateType type, String data) {
         this.type = type;
-        this.data = data;
+        this.data = SplitData(data, 3000);
     }
 
     @Override
     public void fromBytes(ByteBuf buf) {
         this.type = TeamUpdateType.values()[buf.readInt()];
-        data = ByteBufUtils.readUTF8String(buf);
+        int count = buf.readInt();
+        this.data.clear();
+        for (int i = 0; i<count; i++) {
+            this.data.add(ByteBufUtils.readUTF8String(buf));
+        }
     }
 
     @Override
     public void toBytes(ByteBuf buf) {
         buf.writeInt(this.type.ordinal());
-        ByteBufUtils.writeUTF8String(buf, data);
+
+        buf.writeInt(this.data.size());
+        for (String chunk : this.data) {
+            ByteBufUtils.writeUTF8String(buf, chunk);
+        }
 
         //buf.writeInt(this.data.getBytes().length);
         
@@ -49,7 +70,7 @@ public class TeamUpdateMessage implements IMessage {
         }
 
         private void handle(TeamUpdateMessage message, MessageContext ctx) {
-            message.type.update(QuestingData.getQuestingData(HardcoreQuesting.proxy.getPlayer(ctx)).getTeam(), message.data);
+            message.type.update(QuestingData.getQuestingData(HardcoreQuesting.proxy.getPlayer(ctx)).getTeam(), String.join("", message.data));
         }
     }
 }

